pandas/tests/indexing/test_multiindex.py

```diff
@@ -705,6 +705,15 @@ def test_multiindex_symmetric_difference(self):
         result = idx ^ idx2
         assert result.names == [None, None]
 
+    def test_multiindex_contains_dropped(self):
+        # GH 19027
+        idx = MultiIndex.from_product([[1, 2], [3, 4]])
+        assert 2 in idx
+        idx = idx.drop(2)
+        # drop implementation keeps 2 in the levels
+        assert 2 in idx.levels[0]
+        # but it should no longer be in the index itself
```

Comment:
can you test with a non-integer MI as well

Reply:
This comment made me realize that I had focused onto integers specifically when this should apply to all hashable keys (sorry). But, that would mean tuples would then be treated differently from how they currently are, e.g.:
```
In [88]: idx = pd.MultiIndex.from_product([[(1, 2), (2, 3)], [(3, 4), (4, 5)]])

In [89]: ((1, 2), (3, 4)) in idx
Out[89]: True

In [90]: (1, 2) in idx
Out[90]: False

In [91]: ((1, 2),) in idx
Out[91]: True
```
So, should it assume that a nested tuple like in `In [89]` should be treated like tuples are currently treated but a non-nested tuple like in `In [90]` should be treated like strings and ints?

Reply:
Off the same vein, `idx.drop((1, 2))` raises a KeyError.

Reply:
> This comment made me realize that I had focused onto integers specifically when this should apply to all hashable keys (sorry).

I think the idea should just be: any key ``k`` which is not already a tuple must become one (``(k,)``). The result will be, I think, consistent with current behavior (``Out[89]:``, ``Out[90]:``, ``Out[91]:``), and you don't even need to check if key can be hashed - if it can't, it will just fail later on.

Reply:
> Off the same vein, idx.drop((1, 2)) raises a KeyError and might warrant a separate issue

Sure, in principle we could try with ``((1,2),)`` if ``(1,2)`` is not found. However, I personally don't think it's a good idea, as it introduces ambiguity (we should be clear to users that a tuple in a ``MultiIndex`` is first and foremost a key spanning across levels) and increases complexity.

Reply:
Got it, thanks!
Doing it by simply turning non-tuples into tuples like so:
```
if not isinstance(key, tuple):
    try:
        return self.get_loc((key,))
    except (LookupError, TypeError):
        raise KeyError(key)
```
produces a performance warning for slightly larger multi-indeces:
```
In [31]: idx = pd.MultiIndex.from_product([[1, 2], [3, 4]])

In [32]: 2 in idx
Out[32]: True

In [33]: idx = pd.MultiIndex.from_product([[2, 1, 2], [3, 4]])

In [34]: 2 in idx
/home/avi/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/IPython/terminal/ipapp.py:356: PerformanceWarning: indexing past lexsort depth may impact performance.
  self.shell.mainloop()
Out[34]: True
```
Or we could use the fact that a key that has been dropped from a multi-index will return an empty slice when fed into the index's `_get_level_indexer` like so:
```
if not isinstance(key, tuple):
    loc = self._get_level_indexer(key, level=0)
    if isinstance(loc, slice) and loc.start == loc.stop:
        raise KeyError(key)
    return _maybe_to_slice(loc)
```
which does not produce the performance warnings and is ~3 times faster and simply adds to what it is right now:
https://github.com/pandas-dev/pandas/blob/4a8496b2bf02df521b6b7f15843c515545a3d192/pandas/core/indexes/multi.py#L2123-L2125
 but is potentially less clear (and more hacky?). Which one should I use?

Reply:
I ended up using the faster one and added a comment to explain what's going on.

Reply:
Yes, I think it makes sense